Lambdas/GHHooks/EventHandler/Handlers/PRHandler.swift

```diff
@@ -0,0 +1,199 @@
+                return "httpRequestFailed(response: \(response), file: \(file), line: \(line))"
+            case let .tagDoesNotFollowSemVer(release, tag):
+                return "tagDoesNotFollowSemVer(release: \(release), tag: \(tag))"
+            }
+        }
+    }
+
+    let context: HandlerContext
+    let pr: PullRequest
+    let number: Int
+    var event: GHEvent {
+        context.event
+    }
+    var repo: Repository {
+        event.repository
+    }
+    var repoName: String {
+        repo.organization?.name == "vapor" ? repo.name : repo.full_name
+    }
+
+    init(context: HandlerContext) throws {
+        self.context = context
+        self.pr = try context.event.pull_request.requireValue()
+        self.number = try context.event.number.requireValue()
+    }
+
+    func handle() async throws {
+        let action = context.event.action.map({ PullRequest.Action(rawValue: $0) })
+        switch action {
+        case .opened:
+            try await onOpened()
+        case .closed:
+            try await onClosed()
+        default: break
+        }
+    }
+
+    func onOpened() async throws {
+        let creatorName = pr.user.login
+        let creatorLink = pr.user.html_url
+
+        let prLink = pr.html_url
+
+        let body = pr.body == nil ? "" : "\n\n>>> \(pr.body!)".unicodesPrefix(264)
+
+        let description = """
+        ## \(pr.title)
+
+        ### By **[\(creatorName)](\(creatorLink))**
+        \(body)
+        """
+
+        try await context.discordClient.createMessage(
+            channelId: Constants.Channels.issueAndPRs.id,
+            payload: .init(embeds: [.init(
+                title: "[\(repoName)] PR #\(number)".unicodesPrefix(256),
+                description: description,
+                url: prLink,
+                color: .green
+            )])
+        ).guardSuccess()
+    }
+
+    func onClosed() async throws {
+        guard pr.base.ref == "main",
+              let mergedBy = pr.merged_by,
+              let bump = pr.knownLabels.first?.toBump()
+        else { return }
+
+        let previousRelease = try await getLatestRelease()
+
+        let tag = previousRelease.tag_name
+        guard let previousVersion = SemVer(string: tag) else {
+            throw Errors.tagDoesNotFollowSemVer(release: previousRelease, tag: tag)
+        }
+
+        let version = previousVersion.next(bump)
+
+        let acknowledgment: String
+        if pr.user.login == mergedBy.login {
+            acknowledgment = "This patch was authored and released by @\(pr.user.login)."
+        } else {
+            acknowledgment = "This patch was authored by @\(pr.user.login) and released by @\(mergedBy.login)."
+        }
+
+        let release = try await makeNewRelease(version: version, acknowledgment: acknowledgment)
+
+        let ownerLogin = repo.owner.login
+        let repoName = repo.name
+        let url = "https://github.com/\(ownerLogin)/\(repoName)/releases/tag/\(release.tag_name)"
```

Comment:
Ugh. Please at least _consider_ constructing URLs via `URLComponents`, and not just in this one spot? Percent encoding matters, even if in these kinds of GitHub names it generally doesn't become an issue.

Reply:
I copy pasted this from the old bot's code. It's just used for showing to users, not any actual API calls.
I could url encode them all but I don't think you're planning on changing vapor's org name to something weird, or using weird stuff in repo names 😅

Reply:
Changed this url to use the release's `.html_url` since i noticed it has that.

For the record i do always consider url-encoding unless i forget 😅

Reply:
Always best to use what the API provides when it does so, definitely ❤️